luarules/gadgets/unit_builder_priority.lua

```diff
@@ -185,101 +216,223 @@ local function UpdatePassiveBuilders(teamID, interval)
-				if teamStallingMetal then
-					local passivePullMetal = (interval*passiveConsExpense[builderID]['metal']/simSpeed)
-					if passivePullMetal > 0 then
-						local newPullMetal = teamStallingMetal - passivePullMetal
-						if newPullMetal <= 0 then
-							wouldStall = true
-						else
-							teamStallingMetal = newPullMetal
+						
+						if currentBuildSpeed[builderID] ~= wantedBuildSpeed then
+							spSetUnitBuildSpeed(builderID, wantedBuildSpeed)
+							currentBuildSpeed[builderID] = wantedBuildSpeed
+							if DEBUG then spSetUnitRulesParam(builderID, "mybuildspeed", wantedBuildSpeed, LOS_ACCESS) end
 						end
 					end
 				end
 			end
+			if not havePassiveResourcesLeft or j == 2 then 
+				-- Either ran out of resources, or on our second pass
+				roundRobinIndexTeam[teamID] = roundRobinIndex
+				break 
+			else
+				-- We are gonna do a second pass then. 
+				if j == 1 then 
+					roundRobinIndex = 1
+				end
+			end
 		end
+	else
+	-- Special case, we are completely and totally stalled, no resources left for passive builders. 
+		for builderID in pairs(passiveConsTeam) do 
+			if passiveExpense[builderID] then 
+				if currentBuildSpeed[builderID] ~= 0 then
+					spSetUnitBuildSpeed(builderID, 0)
+					currentBuildSpeed[builderID] = 0
+					if DEBUG then spSetUnitRulesParam(builderID, "mybuildspeed", 0, LOS_ACCESS) end
+				end
+			end
+		end	
+	end
+	
+	
+	-- dont remove the resources given to a builder in a round-robin fashion just because they are build target owners
+	-- take them off the buildTargets stack!
+	
+		-- override buildTargetOwners build speeds for a single frame; let them build at a tiny rate to prevent nanoframes from possibly decaying
+	for buildTargetID, builderUnitID in pairs(buildTargets) do 
+		-- if owner is passive, then give it a bit of BP
+		-- TODO: this needs to be smarter
+		-- This ensures that we at least build each unit a tiny bit.
+		if currentBuildSpeed[builderUnitID] <= 0.001 then 
+			-- This builderUnitID has been assigned as passive with no resources
+			spSetUnitBuildSpeed(builderUnitID, max(currentBuildSpeed[builderUnitID], 0.001)) -- 
+			if DEBUG then spSetUnitRulesParam(builderUnitID, "mybuildspeed", max(currentBuildSpeed[builderUnitID], 0.001), LOS_ACCESS) end
+		else
+			-- this builderUnitID has been assigned greater than 0 resources in the round robin pass, so remove it from the next frames clear pass
+			buildTargets[buildTargetID] = nil
+		end
+	end 
+	
+	if VERBOSE then 
+		Spring.Echo(string.format("%d Pstart = %.1f/%.0f Pleft = %.1f/%.0f RRI=%d #passive=%d Stalled=%d", 
+			Spring.GetGameFrame(),
+			passiveMetalStart, passiveEnergyStart,
+			passiveMetalLeft, passiveEnergyLeft, 
+			roundRobinIndex, 
+			numPassiveCons,
+			havePassiveResourcesLeft and 0 or 1
+			)
+		)
+	end
 
+	--- ----------------------------------------- OLD METHOD -------------------------------------------------
+		
+	-- work through passive cons allocating as much expense as we have left
+	--[[
```

Comment:
This stuff can all be deleted - the old method can be found through git history for those who would like to see it 🙂